engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -5631,35 +5640,72 @@ private void checkConcurrentJobsPerDatastoreThreshhold(final StoragePool destPoo
 
         final List<VmWorkJobVO> pendingWorkJobs = _workJobDao.listPendingWorkJobs(
                 VirtualMachine.Type.Instance, vm.getId(),
-                VmWorkAddVmToNetwork.class.getName());
+                VmWorkAddVmToNetwork.class.getName(), network.getUuid());
 
         VmWorkJobVO workJob = null;
         if (pendingWorkJobs != null && pendingWorkJobs.size() > 0) {
-            assert pendingWorkJobs.size() == 1;
+            if (pendingWorkJobs.size() > 1) {
+                s_logger.warn(String.format("The number of jobs to add network %s to vm %s are %d", network.getUuid(), vm.getInstanceName(), pendingWorkJobs.size()));
+            }
             workJob = pendingWorkJobs.get(0);
         } else {
+            if (s_logger.isTraceEnabled()) {
+                s_logger.trace(String.format("no jobs to add network %s for vm %s yet", network, vm));
+            }
 
-            workJob = new VmWorkJobVO(context.getContextId());
+            workJob = createVmWorkJobToAddNetwork(vm, network, requested, context, user, account);
+        }
+        AsyncJobExecutionContext.getCurrentExecutionContext().joinJob(workJob.getId());
 
-            workJob.setDispatcher(VmWorkConstants.VM_WORK_JOB_DISPATCHER);
-            workJob.setCmd(VmWorkAddVmToNetwork.class.getName());
+        if (s_logger.isTraceEnabled()) {
+            s_logger.trace(String.format("job %s to add network %s for vm %s processed", workJob.toString(), network, vm));
+        }
+        if (workJob.getStep() == VmWorkJobVO.Step.Done ||
+            workJob.getStep() == VmWorkJobVO.Step.Error ||
+            workJob.getStep() == VmWorkJobVO.Step.Release) {
+            _workJobDao.expunge(workJob.getId());
+        }
```

Comment:
@weizhouapache the regression that this caused was that the nic couldn't be re-added because the workjob remained in the DB. Are you sure removing it here isn't needed?

Reply:
@DaanHoogland 
this PR lists pending work jobs by vm id and also network uuid.
If you add multiple networks to a vm, cloudstack will create work jobs for all of them, and process the jobs.

Reply:
@DaanHoogland 
I suggest you to try to reproduce the issue #5499 without last commit.

ps: I  found that some vm work jobs are in `Filed` step even if job is finished.

Reply:
yes @weizhouapache I noticed that too.
The scenario from #5499 doen't work, only two nics get added to the VM. I'll keep looking